src/codegen/generators/expression_generator.rs

```diff
@@ -1020,6 +1017,29 @@ impl<'a, 'b> ExpressionCodeGenerator<'a, 'b> {
             })
     }
 
+    pub fn create_llvm_binary_expression_for_pointer(
+        &self,
+        operator: &Operator,
+        left_value: BasicValueEnum<'a>,
+        right_value: BasicValueEnum<'a>,
+    ) -> BasicValueEnum<'a> {
+        let int_type = self.llvm.i32_type();
+
+        let left = if let BasicValueEnum::PointerValue(v) = left_value {
+            v.const_to_int(int_type).as_basic_value_enum()
+        } else {
+            left_value
+        };
+
+        let right = if let BasicValueEnum::PointerValue(v) = right_value {
+            v.const_to_int(int_type).as_basic_value_enum()
+        } else {
+            right_value
+        };
+
+        self.create_llvm_int_binary_expression(operator, left, right)
```

Comment:
I took a look at what C is doing for add/sub operations and it does not call a + but instead calls the gep method (get element pointer)
This ensures the addition always happens with the correct type size so for example int* + 1 will result in jumping one int forward.
Also, mul and div are not allowed which we could add validations for.
Comparisons are still allowed though, and I think these should still be handled like you do here (get the int value of the pointer and compare that)

Reply:
Interestingly, C also does not use the ptrtoint call in comparison, but just loads a pointer, maybe you can also try this?
```
  %36 = load i32*, i32** %3, align 8
  %37 = load i8*, i8** %4, align 8
  %38 = bitcast i8* %37 to i32*
  %39 = icmp ugt i32* %36, %38
  br i1 %39, label %40, label %44

```